modules/artifact-activation/src/main/java/org/mule/runtime/module/artifact/activation/internal/plugin/BundlePluginDependenciesResolver.java

```diff
@@ -0,0 +1,412 @@
+    List<ArtifactPluginDescriptor> resolvedPlugins = new LinkedList<>();
+    List<ArtifactPluginDescriptor> unresolvedPlugins = new LinkedList<>(sortedDescriptors);
+
+    boolean continueResolution = true;
+
+    while (continueResolution) {
+      int initialResolvedCount = resolvedPlugins.size();
+
+      List<ArtifactPluginDescriptor> pendingUnresolvedPlugins = new LinkedList<>();
+
+      for (ArtifactPluginDescriptor unresolvedPlugin : unresolvedPlugins) {
+        if (isResolvedPlugin(unresolvedPlugin, resolvedPlugins)) {
+          sanitizeExportedPackages(unresolvedPlugin, resolvedPlugins);
+          resolvedPlugins.add(unresolvedPlugin);
+        } else {
+          pendingUnresolvedPlugins.add(unresolvedPlugin);
+        }
+      }
+
+      // Will try to resolve the plugins that are still unresolved
+      unresolvedPlugins = pendingUnresolvedPlugins;
+
+      continueResolution = resolvedPlugins.size() > initialResolvedCount;
+    }
+
+    if (unresolvedPlugins.size() != 0) {
+      throw new PluginResolutionError(createResolutionErrorMessage(unresolvedPlugins, resolvedPlugins));
+    }
+
+    return resolvedPlugins;
+  }
+
+  private void verifyPluginExportedPackages(List<ArtifactPluginDescriptor> plugins) {
+    final Map<String, List<String>> exportedPackages = new HashMap<>();
+
+    boolean error = false;
+    for (ArtifactPluginDescriptor plugin : plugins) {
+      for (String packageName : plugin.getClassLoaderModel().getExportedPackages()) {
+        List<String> exportedOn = exportedPackages.get(packageName);
+
+        if (exportedOn == null) {
+          exportedOn = new LinkedList<>();
+          exportedPackages.put(packageName, exportedOn);
+        } else {
+          error = true;
+        }
+        exportedOn.add(plugin.getName());
+      }
+    }
+
+    if (error) {
+      throw new DuplicateExportedPackageException(exportedPackages);
+    }
+  }
+
+  /**
+   * Goes over the elements in the {@code pluginDescriptors} collection looking if it hasn't been resolved yet.
+   *
+   * @param pluginDescriptors plugins to validate.
+   * @param visited           plugins that are already resolved (by either the container or application initially, or by the
+   *                          resolver).
+   * @return the plugins that were obtained initially plus all the ones that were found.
+   */
+  private List<ArtifactPluginDescriptor> getArtifactsWithDependencies(List<ArtifactPluginDescriptor> pluginDescriptors,
+                                                                      Set<BundleDescriptor> visited) {
+    List<ArtifactPluginDescriptor> pluginDescriptorsWithDependences = new ArrayList<>();
+    pluginDescriptorsWithDependences.addAll(pluginDescriptors);
+
+    if (!pluginDescriptors.isEmpty()) {
+      List<ArtifactPluginDescriptor> foundDependencies = new ArrayList<>();
+      pluginDescriptors.stream()
+          .filter(pluginDescriptor -> !pluginDescriptor.getClassLoaderModel().getDependencies().isEmpty())
```

Comment:
why do you need this filter? you are doing a `forEach` after

Reply:
It was like this in the same code present in the deployment module, but it's useless so I'll remove it.